dmscripts/notify_suppliers_of_new_fw_cq_answers.py

```diff
@@ -0,0 +1,81 @@
+from logging import Logger
+from typing import Optional
+from uuid import UUID, uuid4
+
+from dmapiclient import DataAPIClient
+
+from dmutils.email import DMNotifyClient
+from dmutils.email.exceptions import EmailError
+from dmutils.email.helpers import hash_string
+from dmutils.env_helpers import get_web_url_from_stage
+
+
+def notify_suppliers_of_new_fw_cq_answers(
+    data_api_client: DataAPIClient,
+    notify_client: DMNotifyClient,
+    notify_template_id: str,
+    framework_slug: str,
+    stage: str,
+    dry_run: bool,
+    logger: Logger,
+    run_id: Optional[UUID] = None,
+) -> int:
+    run_is_new = bool(run_id)
+    run_id = run_id or uuid4()
+    logger.info(f"{'Starting' if run_is_new else 'Resuming'} run id {{run_id}}", extra={"run_id": str(run_id)})
+
+    framework = data_api_client.get_framework(framework_slug)["frameworks"]
+
+    if framework["status"] != "open":
+        raise ValueError(f"Framework {framework_slug!r} is not open (status is {framework['status']!r})")
+
+    failure_count = 0
+
+    for supplier_framework in data_api_client.find_framework_suppliers_iter(framework_slug):
+        for user in data_api_client.find_users_iter(supplier_id=supplier_framework["supplierId"]):
+            if user["active"]:
+                # generating ref separately so we can exclude certain parameters from the personalization dict
+                notify_ref = notify_client.get_reference(
+                    user["emailAddress"],
+                    notify_template_id,
+                    {
+                        "framework_slug": framework["slug"],
+                        "run_id": str(run_id),
+                    },
+                )
+                if dry_run:
+                    if notify_client.has_been_sent(notify_ref):
```

Comment:
Checking my understanding: this line would return true if we'd made a previous script attempt that contained a failure, and we're doing a `--dry-run` attempt first before we re-run for real?

Reply:
Yeah, so, the `has_been_sent` check is usually done _inside_ the `send_email` method, so we don't have to call it ourselves in the normal case. The only reason we have to call it ourselves is to be able to emit the correct log message for `dry_run` mode.